ebos/eclactionhandler.cc

```diff
@@ -0,0 +1,186 @@
+
+#include <opm/input/eclipse/EclipseState/EclipseState.hpp>
+#include <opm/input/eclipse/Schedule/Schedule.hpp>
+#include <opm/input/eclipse/Schedule/Action/ActionContext.hpp>
+#include <opm/input/eclipse/Schedule/Action/ActionX.hpp>
+#include <opm/input/eclipse/Schedule/Action/State.hpp>
+
+#include <opm/simulators/utils/ParallelSerialization.hpp>
+#include <opm/simulators/wells/BlackoilWellModelGeneric.hpp>
+
+#include <iomanip>
+#include <sstream>
+
+namespace Opm {
+
+EclActionHandler::EclActionHandler(EclipseState& ecl_state,
+                                   Schedule& schedule,
+                                   Action::State& actionState,
+                                   SummaryState& summaryState,
+                                   BlackoilWellModelGeneric& wellModel,
+                                   Parallel::Communication comm)
+    : ecl_state_(ecl_state)
+    , schedule_(schedule)
+    , actionState_(actionState)
+    , summaryState_(summaryState)
+    , wellModel_(wellModel)
+    , comm_(comm)
+{
+}
+
+void EclActionHandler::applyActions(int reportStep,
+                                    double sim_time,
+                                    const TransFunc& transUp)
+{
+    const auto& actions = schedule_[reportStep].actions();
+    if (actions.empty())
+        return;
+
+    Action::Context context( summaryState_, schedule_[reportStep].wlist_manager() );
+    auto now = TimeStampUTC( schedule_.getStartTime() ) + std::chrono::duration<double>(sim_time);
+    std::string ts;
+    {
+        std::ostringstream os;
+        os << std::setw(4) <<                      std::to_string(now.year())  << '/'
+           << std::setw(2) << std::setfill('0') << std::to_string(now.month()) << '/'
+           << std::setw(2) << std::setfill('0') << std::to_string(now.day()) << "  report:" << std::to_string(reportStep);
+
+        ts = os.str();
+    }
+
+    bool commit_wellstate = false;
+    for (const auto& pyaction : actions.pending_python(actionState_)) {
+        auto sim_update = schedule_.runPyAction(reportStep, *pyaction, actionState_,
+                                                ecl_state_, summaryState_);
+        this->applySimulatorUpdate(reportStep, sim_update, commit_wellstate, transUp);
+    }
+
+    auto simTime = asTimeT(now);
+    for (const auto& action : actions.pending(actionState_, simTime)) {
+        auto actionResult = action->eval(context);
+        if (actionResult) {
+            std::string wells_string;
+            const auto& matching_wells = actionResult.wells();
+            if (!matching_wells.empty()) {
+                for (std::size_t iw = 0; iw < matching_wells.size() - 1; iw++)
+                    wells_string += matching_wells[iw] + ", ";
+                wells_string += matching_wells.back();
+            }
+            std::string msg = "The action: " + action->name() + " evaluated to true at " + ts + " wells: " + wells_string;
+            OpmLog::info(msg);
```

Comment:
This block can be simplified quite considerably if we switch to using libfmt instead:
```C++
const auto msg =
    fmt::format("The action {} evaluated to true at {}. Well{}: {}",
                action->name(), ts,
                matching_wells.size() != 1 ? "s" : "",
                fmt::join(matching_wells, ", "));
OpmLog::info(msg)
```

Reply:
i guess i can do it. we'd get wells for an empty matching wells, not sure if the original if carried any meaning though.

Reply:
> we'd get wells for an empty matching wells

I'm not sure I completely understand what you're saying here.  The original version would print
```
" wells: "
```
while the libfmt-based version prints
```
". Wells: "
````
in that case (empty `matching_wells` array).  That's a minor difference in my opinion.

Reply:
yeah, sorry I was rather unclear. my point was that IF we can be here with matching_wells.empty(), maybe we should consider not printing the last part. the original code was just as broken in this aspect.

Reply:
ie something like
```
    std::string wmsg;
    if (!matching_wells.empty())
            wmsg = fmt::format("Well{}: {}",
                                             matching_wells.size() != 1 ? "s" : "",
                                             fmt::join(matching_wells, ", "));
            const auto msg =
                    fmt::format("The action {} evaluated to true at {}. {}",
                                action->name(), ts, wmsg);
```

Reply:
> ie something like

Good!  I like it.  Let's do that.  The next step might be to use libfmt to format the `ts` too as doing so will (easily) allow us to output date strings of the form
```
15-Aug-2022
```
instead of the current format
```
2022/08/15
```
I can submit a follow-up PR to that effect unless you want to do it.

Reply:
as this smells like a custom formatter i'm not entirely sure where is at, please do a follow-up.